src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityContext.cs

```diff
@@ -0,0 +1,178 @@
+			SpecialType.System_Enum,
+			SpecialType.System_Boolean,
+			SpecialType.System_Char,
+			SpecialType.System_SByte,
+			SpecialType.System_Byte,
+			SpecialType.System_Int16,
+			SpecialType.System_UInt16,
+			SpecialType.System_Int32,
+			SpecialType.System_UInt32,
+			SpecialType.System_Int64,
+			SpecialType.System_UInt64,
+			SpecialType.System_Decimal,
+			SpecialType.System_Single,
+			SpecialType.System_Double,
+			SpecialType.System_String,
+			SpecialType.System_IntPtr,
+			SpecialType.System_UIntPtr,
+			SpecialType.System_ValueType
+		);
+
+		public ImmutabilityContext( IEnumerable<ImmutableTypeInfo> extraImmutableTypes ) {
+			m_extraImmutableTypes = extraImmutableTypes
+				.ToImmutableDictionary(
+					info => info.Type,
+					info => info
+				);
+		}
+
+		/// <summary>
+		/// Determines if a type is known to be immutable.
+		/// </summary>
+		/// <param name="type">The type to check</param>
+		/// <param name="kind">The degree of immutability</param>
+		/// <param name="diag">If this method returns false, an explaination for why its not known to be immutable.</param>
+		/// <returns>Is the type immutable?</returns>
+		public bool IsImmutable(
+			ITypeSymbol type,
+			ImmutableTypeKind kind,
+			Location location,
+			out Diagnostic diagnostic
+		) {
+			if ( kind == ImmutableTypeKind.None ) {
+				throw new ArgumentException(
+					"ImmutabilityKind.None is not a valid question to ask this function",
+					nameof( kind )
+				);
+			}
+
+			diagnostic = null;
+
+			// Things like int are totally OK
+			if ( m_totallyImmutableSpecialTypes.Contains( type.SpecialType ) ) {
+				return true;
+			}
+
+			// "new object()" are always immutable (and that's the only
+			// constructor for System.Object) but in general things of type
+			// System.Object (any reference type) may be mutable.
+			//
+			// This is hard-coded (rather than in m_extraImmutableTypes) to
+			// avoid the ITypeSymbol lookup.
+			if ( kind == ImmutableTypeKind.Instance && type.SpecialType == SpecialType.System_Object ) {
+				return true;
+			}
+
+			// Check for [Immutable] etc.
+			if ( GetImmutabilityFromAttributes( type ).HasFlag( kind ) ) {
+				return true;
+			}
+
+			// Check if we were otherwise told that this type is (sufficiently)
+			// immutable.
+			if ( m_extraImmutableTypes.TryGetValue( type, out var info ) && info.Kind.HasFlag( kind ) ) {
+				return true;
+			}
+
+			switch( type.TypeKind ) {
+				case TypeKind.Error:
+					// Just say this is fine -- there is some other compiler
+					// error in this case and we don't need to pile on.
+					return true;
+
+				case TypeKind.Enum:
+					// Enums are like ints -- always immutable
+					return true;
+
+				case TypeKind.Array:
+					diagnostic = Diagnostic.Create(
+						Diagnostics.ArraysAreMutable,
+						location,
+						type.Name
+					);
+
+					return false;
+
+				case TypeKind.Delegate:
+					// Delegate members just give a name to a function type.
+					return true;

```

Comment:
This is a difference from the current analyzer. Seems fine to me?

Reply:
Would this change allow something like this to pass the analyzer?

```cs
public delegate string SillyDelegate( string key );

[Immutable]
public class Foo {
    public SillyDelegate m_func;

    public Foo( Dictionary<string, string> sillyCache, IConfigViewer viewer ) {
        m_func = key => {
			if( !sillyCache.TryGetValue( key, out string value ) ) {
				sillyCache[key] = viewer.GetValue( key );
			}
			return value;
        };
    }

	public string GetValue( string key ) => m_func( key );
}
```

Reply:
Yes, this is wrong:

```cs
const string src = @"
	class X {
		delegate void Y();
		public Y m_y;
		event Y m_z;
	}";

var tree = CSharpSyntaxTree.ParseText( src );
var compilation = CSharpCompilation.Create( "assembly", new [] { tree } );
var model = compilation.GetSemanticModel( tree );
var decl = tree.GetRoot().ChildNodes().First();
var xtype = (INamedTypeSymbol)model.GetDeclaredSymbol( decl );
var members = xtype.GetMembers();

foreach( var member in members ) {
	var memberTypeInfo =
		member is IPropertySymbol prop ? $" of type {prop.Type.Name} which is a {prop.Type.TypeKind}"
		: member is IFieldSymbol field ? $" of type {field.Type.Name} which is a {field.Type.TypeKind}"
		: null;

	Console.WriteLine( $"{member.Name} is a {member.Kind}{memberTypeInfo} (implicit: {member.IsImplicitlyDeclared})" );
}
```

outputs:

```
Y is a NamedType (implicit: False)
m_y is a Field of type Y which is a Delegate (implicit: False)
add_m_z is a Method (implicit: True)
remove_m_z is a Method (implicit: True)
m_z is a Event (implicit: False)
.ctor is a Method (implicit: True)
```

This code will get used to look at a fields type, and the only time we see TypeKind == delegate is when we have a field/property that uses a delegate, not when a delegate is declared (that one is just a NamedType member)... I got mixed up.

Reply:
Fixed; thanks.